predict.py

```diff
@@ -1,94 +1,280 @@
+
 
-def download_weights(url, dest):
-    start = time.time()
-    print("downloading url: ", url)
-    print("downloading to: ", dest)
-    subprocess.check_call(["pget", "-xf", url, dest], close_fds=False)
-    print("downloading took: ", time.time() - start)
+class Predictor:
 
-class Predictor(BasePredictor):
     def setup(self) -> None:
-        # Download weights
-        if not os.path.exists(MODEL_CACHE):
-            download_weights(MODEL_URL, MODEL_CACHE)
+        """
+        setup the predictor and server
+        """
+
+        # when the endpoint http://<host>:<port>/predictions/<prediction_id>/cancel is called
+        # (user presses the cancel button on the website)
+        # cog will send SIGUSR1 to the process that is running predict. When this happens we must 
+        # terminate the current prediction and return from predict as soon as possible.
+        # this signal handler passes along the cancellation to the multigpu inference server 
+        signal.signal(signal.SIGUSR1, self._get_signal_handler())
+
+        # stuff that the server needs to know
+        self.WORLD_SIZE = torch.cuda.device_count()
+        self.MODEL_CACHE = "checkpoints"
+        self.MODEL_URL = "https://weights.replicate.delivery/default/hunyuanvideo-community/HunyuanVideo/model.tar"
+        self.HTTP_PORT = 5001
+        print(f"inference server will use {self.WORLD_SIZE} devices")
+
+        if not os.path.exists(self.MODEL_CACHE):
+            self._download_weights()
+        
+        # initialize the model specific predictor and server
+        # the predictor is a wrapper around the HunyuanVideoPipeline
+        # the server is an inference server that runs the predictor
+        self.model = HunyuanPredictor(self.MODEL_CACHE, self.WORLD_SIZE, self.HTTP_PORT, self.WORLD_SIZE)
+        self.server_process = mp.Process(target=run_server, args=(self.model, self.HTTP_PORT, self.WORLD_SIZE))
+        self.server_process.start()
+        self.request_cancelled = False # internal flag that is used to manage cancellation of inference requests
         
-        self.transformer = HunyuanVideoTransformer3DModel.from_pretrained(
-            MODEL_CACHE,
-            subfolder="transformer",
-            torch_dtype=torch.bfloat16,
-        )
-        self.pipe = HunyuanVideoPipeline.from_pretrained(
-            MODEL_CACHE,
-            transformer=self.transformer,
-            torch_dtype=torch.float16,
-        )
-        self.pipe.vae.enable_tiling()
-        self.pipe.to("cuda")
+        # warmup
+        predict_args = {
+            'prompt': 'A cat walks on the grass, realistic style',
+            'width': 864,
+            'height': 480,
+            'video_length': 61,
+            'infer_steps': 1,
+            'embedded_guidance_scale': 6.0,
+            'fps': 24,
+            'seed': 42,
+            'output_path': 'warmup.mp4'
+        }
+        self._request_and_wait(predict_args, timeout=float('inf'))
```

Comment:
probably fine to rely on cog's setup timeout here but might be worth setting to ~an hour instead of `inf`.